osquery/tables/system/darwin/gatekeeper.cpp

```diff
@@ -0,0 +1,172 @@
+
+const std::string kGkeStatusPath{
+    "/var/db/SystemPolicy-prefs.plist"};
+
+const std::string kGkeBundlePath{
+    "/var/db/gke.bundle/Contents/version.plist"};
+
+const std::string kGkeOpaquePath{
+    "/var/db/gkopaque.bundle/Contents/version.plist"};
+
+const std::string kPolicyDb = "/var/db/SystemPolicy";
+
+bool isGateKeeperDevIdEnabled() {
+  sqlite3* db = nullptr;
+  auto rc = sqlite3_open_v2(
+      kPolicyDb.c_str(),
+      &db,
+      (SQLITE_OPEN_READONLY | SQLITE_OPEN_PRIVATECACHE | SQLITE_OPEN_NOMUTEX),
+      nullptr);
+  if (rc != SQLITE_OK || db == nullptr) {
+    VLOG(1) << "Cannot open Gatekeeper DB: " << rc << " "
+            << getStringForSQLiteReturnCode(rc);
+    if (db != nullptr) {
+      free(db);
+    }
+  }
+
+  std::string query = "SELECT disabled FROM authority WHERE label = 'Developer ID'";
+  sqlite3_stmt* stmt = nullptr;
+  rc = sqlite3_prepare_v2(db, query.c_str(), -1, &stmt, nullptr);
+
+  while ((rc = sqlite3_step(stmt)) == SQLITE_ROW) {
+    int value = sqlite3_column_int(stmt, 0);
+    if (value == 1) {
+      // Clean up.
+      sqlite3_finalize(stmt);
+      free(db);
+      return false; // return false if any rows say "disabled"
+    }
+  }
+  sqlite3_finalize(stmt);
+  free(db);
+  return true;
+}
+
+QueryData genGateKeeper(QueryContext& context) {
+  Row r;
+
+  auto gke_status = SQL::selectAllFrom("plist", "path", EQUALS, kGkeStatusPath);
+
+  if (gke_status.empty()) {
+    r["assessments_enabled"] = INTEGER(0);
+  }
+
+  for (const auto& row : gke_status) {
+    if (row.at("key") == "enabled" && row.at("value") == "yes" ) {
+      r["assessments_enabled"] = INTEGER(1);
+      if (isGateKeeperDevIdEnabled()) {
+        r["dev_id_enabled"] = INTEGER(1);
+      } else {
+        r["dev_id_enabled"] = INTEGER(0);
+      }
+    } else {
+      r["assessments_enabled"] = INTEGER(0);
+      r["dev_id_enabled"] = INTEGER(0);
+    }
+  }
+
+  auto gke_bundle = SQL::selectAllFrom("plist", "path", EQUALS, kGkeBundlePath);
+
+  if (gke_bundle.empty()) {
+    r["version"] = std::string();
+  }
+
+  for (const auto& row : gke_bundle) {
+    if (row.at("key") == "CFBundleShortVersionString") {
+      r["version"] = row.at("value");
+    }
+  }
+
+  auto gke_opaque = SQL::selectAllFrom("plist", "path", EQUALS, kGkeOpaquePath);
+
+  if (gke_opaque.empty()) {
+    r["opaque_version"] = std::string();
+  }
+
+  for (const auto& row : gke_opaque) {
+    if (row.at("key") == "CFBundleShortVersionString") {
+      r["opaque_version"] = row.at("value");
+    }
+  }
+  return {r};
+}
+
+void genGateKeeperApprovedAppRow(sqlite3_stmt* stmt, Row& r) {
+  for (int i = 0; i < sqlite3_column_count(stmt); i++) {
+    auto column_name = std::string(sqlite3_column_name(stmt, i));
+    auto column_type = sqlite3_column_type(stmt, i);
+    if (column_type == SQLITE_TEXT) {
+      auto value = sqlite3_column_text(stmt, i);
+      if (value != nullptr) {
+        r[column_name] = std::string((const char*)value);
```

Comment:
@zwass do you know if it'd be possible to pull assignments out, rearrange some things and avoid the `reinterpret_cast`? We try to avoid these if at all possible.